stan/math/prim/mat/fun/offset_multiplier_constrain.hpp

```diff
@@ -0,0 +1,116 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_OFFSET_MULTIPLIER_CONSTRAIN_HPP
+#define STAN_MATH_PRIM_MAT_FUN_OFFSET_MULTIPLIER_CONSTRAIN_HPP
+
+#include <stan/math/prim/mat/err/check_cholesky_factor.hpp>
+#include <stan/math/prim/mat/err/check_square.hpp>
+#include <stan/math/prim/mat/err/check_finite.hpp>
+#include <stan/math/prim/mat/fun/log.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/err/check_consistent_size.hpp>
+#include <stan/math/prim/scal/fun/identity_constrain.hpp>
+#include <stan/math/prim/mat/fun/diagonal.hpp>
+#include <cmath>
+#include <limits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the linearly transformed value for the specified unconstrained input
+ * and specified offset and multiplier.
+ *
+ * <p>The transform applied is
+ *
+ * <p>\f$f(x) = mu + sigma * x\f$
+ *
+ * <p>where mu is the offset and sigma is the multiplier.
+ *
+ * <p>If the offset is zero and the multiplier is one this
+ * reduces to <code>identity_constrain(x)</code>.
+ *
+ * @tparam T type of vector
+ * @tparam M type of offset
+ * @tparam S type of multiplier
+ * @param[in] x Unconstrained vector input
+ * @param[in] mu offset of constrained output
+ * @param[in] sigma multiplier of constrained output
+ * @return linear transformed value correspdonding to inputs
+ * @throw std::domain_error if sigma <= 0
+ * @throw std::domain_error if mu is not finite
+ */
+template <typename T, typename M, typename S>
+inline Eigen::Matrix<typename return_type<T, M, S>::type, -1, 1>
+offset_multiplier_constrain(const Eigen::Matrix<T, -1, 1>& x,
+                            const Eigen::Matrix<M, -1, 1>& mu,
+                            const Eigen::Matrix<S, -1, -1>& sigma) {
+  static const char* function = "offset_multiplier_constrain";
+  check_finite(function, "uncontrained vector", x);
+  check_finite(function, "offset", mu);
+  check_finite(function, "multiplier", sigma);
+  check_cholesky_factor(function, "multiplier", sigma);
+  check_square(function, "multiplier", sigma);
+  const size_t N = sigma.col(0).size();
+  check_consistent_size(function, "uncontrained vector", x, N);
+  check_consistent_size(function, "offset", mu, N);
+  if (sigma == Eigen::Matrix<S, -1, -1>::Identity(N, N)) {
```

Comment:
This looks pretty inefficient; won't this create an `N x N` matrix that's of type `S`? If that's `stan::math::var`, we'll just be tossing a bunch of stuff on the stack, no?

Maybe it's worth creating an `is_identity()` or `is_identity_matrix()` function for this purpose?

Reply:
You're right. The doubles should be pulled out in that case. I wonder if it would make sense to just always perform `mu + sigma * x`? Basically, you should not be calling this function in the case that `mu` is zero and `sigma` is identity anyway. And in the case where either `mu=0` or `sigma=id` it might not matter so much compared to pulling the doubles out and checking whether it's zero or the identity. What do you think? (I suspect you have a better handle on C++ efficiency than I do.) Thanks!

Reply:
1) I bet it's almost as cheap to just let Eigen handle it (e.g. I bet it notices when elements are zero). We can test with a [microbenchmark](https://github.com/seantalts/perf-math)
2) We probably don't want to actually call identity_constrain -there's always some chance the C++ compiler can't tell that that's a no-op

Reply:
> won't this create an N x N matrix that's of type S

It should just create an expression template, which is relatively cheap.

> I bet it's almost as cheap to just let Eigen handle it (e.g. I bet it notices when elements are zero). We can test with a microbenchmark

I'll bet against.  Checks for zero are super expensive compared to just multiplying by zero when the values are `double`.  It's only with autodiff that the costs radically change.

> We probably don't want to actually call identity_constrain -there's always some chance the C++ compiler can't tell that that's a no-op

If it's inline and no-op, I think that chance is very low.  Pretty much the entire Boost and Eigen libraries rely on efficient inlining in this way.  You can always try it.

Reply:
> It should just create an expression template, which is relatively cheap.

@bob-carpenter, it's not as cheap as you think. The short is that the expression template alone doesn't pop anything to the stack, but when we check for equality, we must be instantiating a matrix. 

Here's a quick test (I called it `test/unit/Eigen_template_expression_test.cpp`). I instantiated a 25x25 matrix and only added 2 elements to the autodiff stack, `zero` and `one`. After comparison to the Eigne template expression, we end up with 627 elements on the stack (the 2 original elements + 25x25 new ones):
```
#include <gtest/gtest.h>
#include <stan/math/rev/mat.hpp>

TEST(EigenTemplateExpression, Identity) {
  ASSERT_EQ(0, stan::math::ChainableStack::instance().var_stack_.size());
  int N = 25;
  stan::math::var zero = 0, one = 1;
  Eigen::Matrix<stan::math::var, -1, -1> x(N, N);
  for (int i = 0; i < N; ++i) {
    for (int j = i; j < N; ++j) {
      if (i == j)
	x(i, j) = one;
      else {
	x(i, j) = zero;
	x(j, i) = zero;
      }
    }
  }
      
  ASSERT_EQ(2, stan::math::ChainableStack::instance().var_stack_.size());
  Eigen::Matrix<stan::math::var, -1, -1>::Identity(N, N);
  EXPECT_EQ(2, stan::math::ChainableStack::instance().var_stack_.size())
    << "the template expression doesn't add anything to the stack";
    
  bool is_identity = (x == Eigen::Matrix<stan::math::var, -1, -1>::Identity(N, N));
  EXPECT_EQ(N * N + 2, stan::math::ChainableStack::instance().var_stack_.size());
}
```

Reply:
Here is a benchmark showing all three versions for all doubles or all vars. You can edit it to find all combinations if you like. https://github.com/seantalts/perf-math/blob/master/checks_vs_mult.cpp

Here are the results:
```
Seans-MBP ~/scm/perf-math (master) $ make checks_vs_mult && ./checks_vs_mult 
c++ -Imath/lib/eigen_3.3.3/ -Ibenchmark/include -std=c++1y -Imath/ -Imath/lib/boost_1.66.0 -O3  -Lbenchmark/build/src  checks_vs_mult.cpp  -lbenchmark -o checks_vs_mult
2019-01-25 12:23:32
Running ./checks_vs_mult
Run on (8 X 2900 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 262K (x4)
  L3 Unified 8388K (x1)
----------------------------------------------------------------------------------------------------------
Benchmark                                                                   Time           CPU Iterations
----------------------------------------------------------------------------------------------------------
BM_Check<double, double, double>/25                                       533 ns        532 ns    1284970
BM_BobCheck<double, double, double>/25                                    791 ns        790 ns     880260
BM_Mult<double, double, double>/25                                        493 ns        493 ns    1419930
BM_Check<stan::math::var, stan::math::var, stan::math::var>/25          18055 ns      18052 ns      37993
BM_BobCheck<stan::math::var, stan::math::var, stan::math::var>/25         802 ns        802 ns     924263
BM_Mult<stan::math::var, stan::math::var, stan::math::var>/25           11971 ns      11966 ns      78520
BM_Check<double, double, double>/500                                   201351 ns     201322 ns       3309
BM_BobCheck<double, double, double>/500                                297666 ns     297540 ns       2200
BM_Mult<double, double, double>/500                                    164365 ns     164354 ns       4102
BM_Check<stan::math::var, stan::math::var, stan::math::var>/500       9852427 ns    9667372 ns        121
BM_BobCheck<stan::math::var, stan::math::var, stan::math::var>/500     478972 ns     478466 ns       1465
BM_Mult<stan::math::var, stan::math::var, stan::math::var>/500        6893068 ns    6880771 ns        105
```

Reply:
So yeah, @syclik is right that these instantiations kill the times, I was right that the plain multiplications are faster, and Bob is right that writing our own is_identity and is_zero checks works best if we're dealing with `var` matrices (but not for `double`).